### PR TITLE
#6714 esr reconciliation + refactoring

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/6714

### CODE REVIEW

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java [hunk 1]
```diff
@@ -1282,8 +1316,10 @@ public void linkBankStatementLinesByPaymentIds(@NonNull final Map<PaymentId, Ban
 
 		final Set<PaymentId> paymentIds = bankStatementLineRefIdIndexByPaymentId.keySet();
 
+		final List<I_ESR_ImportLine> esrImportLines = esrImportDAO.retrieveLines(paymentIds);
+
 		final ImmutableMap<PaymentId, I_ESR_ImportLine> paySelectionLinesByPaymentId = Maps.uniqueIndex(
-				esrImportDAO.retrieveLines(paymentIds),
+				esrImportLines,
 				esrImportLine -> PaymentId.ofRepoId(esrImportLine.getC_Payment_ID()));
```
Comment (metas-ts) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java:1323]: could we run into multiple esrImportLines having the same C_Payment_ID? or are we ruling this out, e.g. with aunique contraint?
(or maybe the answer is in the code above?)

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java [hunk 2]
```diff
@@ -1294,6 +1330,46 @@ public void linkBankStatementLinesByPaymentIds(@NonNull final Map<PaymentId, Ban
...
+	@VisibleForTesting
+	void updateESRImportReconciledStatus(@NonNull final Set<ESRImportId> esrImportIds)
+	{
+		if (esrImportIds.isEmpty())
+		{
+			// shall NOT happen
+			return;
+		}
+
+		final ImmutableSet<ESRImportId> notReconciledESRImportIds = Services.get(IQueryBL.class)
```
Comment (metas-ts) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java:1354]: i think it's cleaner if you move this method into esrImportDAO..

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java:1323]: Yes, there's a unique index: 

```sql


-- DROP INDEX public.esr_importline_payment_uq;

CREATE UNIQUE INDEX esr_importline_payment_uq
  ON public.esr_importline
  USING btree
  (c_payment_id)
  WHERE c_payment_id IS NOT NULL;

```

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRImportBL.java]: Moved

**Pull request merged**